sidekick/lib/src/util/directory_extension.dart

```diff
@@ -0,0 +1,21 @@
+import 'dart:io';
+
+import 'package:dartx/dartx_io.dart';
+import 'package:path/path.dart';
+
+extension DirectoryExtension on Directory {
+  bool isWithinOrEqual(Directory dir) {
+    return this.isWithin(dir) ||
+        // canonicalize is necessary, otherwise '/a/b/c' != '/a/b/c/' != '/a/b/c/.' != '/a/b/c/../c'
+        dir.canonicalized.path == canonicalized.path;
+  }
+
+  /// When [path] is absolute, returns the directory at that path.
+  /// Else resolves the [path] relative to this directory.
+  Directory resolveAbsoluteOrRelativeDirPath(String path) =>
+      (Directory(path).isAbsolute ? Directory(path) : directory(path))
+          .canonicalized;
```

Comment:
`canonicalize` always converts paths to absolute paths. This means this method works not different from directly calling `canonicalize`

https://github.com/dart-lang/path/blob/9955b27b9bb98d87591208e19eb01c51d29fd467/lib/src/context.dart#L344-L351

Reply:
If a relative path is given to `canonicalize`, it resolves it relative to the current working directory. This method resolves it relative to the directory it is called on - not necessarily on the current working directory.

```dart
    final resolveFromDir = Directory('/a/b/c');

    // path is relative
    print(resolveFromDir.resolveAbsoluteOrRelativeDirPath('d/e/f').path);
          // -> /a/b/c/d/e/f
    print(canonicalize('d/e/f'));
          // -> <Directory.current.path>/d/e/f

    // path is absolute
    print(resolveFromDir.resolveAbsoluteOrRelativeDirPath('/d/e/f').path);
          // -> /d/e/f
    print(canonicalize('/d/e/f'));
          // d/e/f
```

But I found `Context(current: resolveRelativeToThisDir).canonicalize` which achieves the same as this method!